### PR TITLE
Country dataset

### PR DESCRIPTION
**Pull request by puhuk**
cc @pmeier

### CODE REVIEW

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:21]: This is missing the `"valid"` split.

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:45]: We should have an extra `self._check_exists()` here and raise an error, if it comes back negative. Otherwise the user will probably get an uninformative error while setting `download=False` and the data is not present.

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:27]:
```suggestion
    _URL = "https://openaipublic.azureedge.net/clip/data/country211.tgz"
    _MD5 = "84988d7644798601126c29e9877aab6a"
```

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:73]:
```suggestion
        download_and_extract_archive(self._URL, download_root=self.root, md5=self._MD5)
```

**Review by pmeier — COMMENTED** [review]
Hey @puhuk, thanks a lot for the PR. I've added a few comments inline. There are three more issues:

1. You used `find_classes` and `make_dataset` to create the dataset. This is far better than doing it yourself, but it would be even better to simply subclass `ImageFolderDataset` so you don't have to worry about that at all. Basically, you only need to handle the input parsing and download and simply pass `os.path.join(root, "country", self._split)` to `super().__init__()` (along with the transforms of course). If you do that, you don't need to define `__len__` and `__getitem__` at all.
2. Could you add the dataset to our documentation? For that, simply add the class name [to the list in `docs/source/datasets.rst`](https://raw.githubusercontent.com/pytorch/vision/main/docs/source/datasets.rst).
3. Could you add tests for the dataset? You can create a `Country211TestCase` in [this file](https://github.com/pytorch/vision/blob/e65a857b5487a8493bc8a80a95d64d9f049de347/test/test_datasets.py#L2206). You basically only need to provide a minimal configuration as well as provide fake data. You can read [this docstring](https://github.com/pytorch/vision/blob/e65a857b5487a8493bc8a80a95d64d9f049de347/test/datasets_utils.py#L186) or look at the other test cases for examples. Due to the very easy structure, all the fake data can be created by `datasets_utils.create_image_folder`.

File: torchvision/datasets/country211.py
Comment (puhuk) [reply torchvision/datasets/country211.py:73]: @pmeier Thank you for your kind review. Let me check and correct it soon :)

**Comment by puhuk** [comment]
@pmeier Thank you for your kind review. Let me check and correct it soon :)

**Comment by puhuk** [comment]
@pmeier  I resend PR with your suggestion and test code.
But there are errors, could you advise me about how to resolve this.

File: ClassName.txt
Comment (pmeier) [inline ClassName.txt:1]: This file shouldn't be here, please remove it.

File: docs/source/datasets.rst
Comment (pmeier) [inline docs/source/datasets.rst:76]: This list should alphabetically sorted. Please put it in the right place.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:2210]: This is the default for `ImageDatasetTestCase` so we don't need to set it at all.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:2231]: 1. As is, this will create images with the following path `{tmpdir}/country211/{cls}/{cls}`. So the class name occurs twice, while the split does not occur at all. 
2. The file name does not reflect the actual data. If you want it to be accurate it should be something like `f"{:7d}_{:.6f}_{:.6f}"` with all three randomly generated. On the other hand, the file name is irrelevant for the dataset. Thus, I suggest to go with the simplest choice and only append `.jpg` to the index.

```suggestion
        split_folder = pathlib.Path(tmpdir) / "country211" / config["split"]
        split_folder.mkdir(parents=True, exist_ok=True)

        return sum(
            [
                len(
                    datasets_utils.create_image_folder(
                        split_folder,
                        name=cls,
                        file_name_fn=lambda idx: f"{idx}.jpg",
                        num_examples=5,
                    )
                )
                for cls in ("AD", "BS", "GR")
            ]
        )
```

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:2249]: This seems copy-pasted from another dataset. These files do not exist for the Country211 dataset.

```suggestion
```

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:72]: 1. There is no `self._images_folder`
2. It is sufficient to only check `self._base_folder` since it is inside `self.root`, i.e. if it exists the parent also have to exist

```suggestion
        return self._base_folder.exists() and self._base_folder.is_dir()
```

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:55]: `self._image_files` does not exist

```suggestion
        return len(self.samples)
```

**Review by pmeier — CHANGES_REQUESTED** [review]
@puhuk I suggested some changes inline. After that, my comment for inheriting from `ImageFolderDataset` (https://github.com/pytorch/vision/pull/5138#pullrequestreview-842564390) is still open.

Don't worry about the merge conflicts yet, we can fix them before merge. Plus, there are some unrelated failing test that you also don't need to worry about.

File: test/test_datasets.py
Comment (puhuk) [reply test/test_datasets.py:2249]: I mistakenly forgot to delete those lines.
Thanks for the review 👍

**Comment by puhuk** [comment]
@pmeier I resend the PR with your review.
But I don't exactly understand how to implement by inheriting ``ImageFolderDataset``.
May I ask for your help with this.

File: test/test_datasets.py
Comment (pmeier) [inline test/test_datasets.py:2468]:
```suggestion
```

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:13]: This description is copy-pasted from another dataset. Please fix it.

File: torchvision/datasets/country211.py
Comment (pmeier) [inline torchvision/datasets/country211.py:33]: There is no mention of the download flag in the docstring. Please add it.

**Review by pmeier — COMMENTED** [review]
@puhuk I pushed the changes to inherit from `ImageFolder`. There are two things left to fix in the docstring. Afterwards this is good to go.

File: torchvision/datasets/country211.py
Comment (pmeier) [reply torchvision/datasets/country211.py:13]: My bad, I was confused by the amount of dataset PRs that I review ATM.

**Review by NicolasHug — APPROVED** [review]
Thanks a lot @puhuk for the PR and @pmeier for the review!

In 9eb4f11 I pushed some minor changes to address @pmeier 's last comments, and to make the tests slightly more robust. LGTM, I'll merge when green. Thanks again!

**Comment by puhuk** [comment]
@pmeier Really thanks to your review and re-creation of my humble code.
Hope to contribute with better code in next time.

**Pull request merged**
